README.md
# shinemas2R

`shinemas2R` is a R package that  queries the database SHiNeMaS ([Seeds History and Network Management System](https://sourcesup.renater.fr/frs/?group_id=2295)) and format data for specific R packages.

Note that an old verson of shinemas2R code is [here](https://github.com/priviere/shinemas2R_deprecated)

## Install

```r
library(devtools)
devtools::install_github("priviere/shinemas2R")
```

## Usage

`shinemas2R` has one function `shinemas()` which queries SHiNeMaS and formats data for specific R packages.
The query is based on a webservice developped within SHiNeMaS v2 by Laetitia Courgey and Yannick de Oliviera. 
The following information are needed to connect to the database:  db_url, user, password and token.
The token can be taken through the web interface of SHiNeMaS by cliking "Get Token" on the top left side of the page.

## Packages supported

### PPBstats

[PPBstats](https://priviere.github.io/PPBstats_web_site/) is a R package for Participatory Plant Breeding statisticial analyses. 

Four queries are supported:

- `PPBstats_data_network_unipart_seed_lots`, for network analysis on
[unipart network for seed lots analysis](https://priviere.github.io/PPBstats_web_site/book/unipart-network-for-seed-lots-analysis.html#format-the-data), 
[unipart network for location analysis](https://priviere.github.io/PPBstats_web_site/book/unipart-network-for-location-analysis.html#format-the-data-1) and 
[bipart network analysis](https://priviere.github.io/PPBstats_web_site/book/bipart-network-analysis.html#format-the-data-2).
- `PPBstats_data_agro`, for [agronomic analysis](https://priviere.github.io/PPBstats_web_site/book/intro-agro.html#data-agro)
- `PPBstats_data_agro_SR`, for [response to selection analysis](https://priviere.github.io/PPBstats_web_site/book/family-4.html#response-to-selection)
- `PPBstats_data_agro_HA`, for [local adaptation analysis](https://priviere.github.io/PPBstats_web_site/book/family-4.html#study-local-adaptation)


## Examples


For example, first get the formated data set
```r
library(shinemas2R)
data_agro = shinemas2R::shinemas(
  db_url = "shinemas.local.seed",
  user = "wheat",
  password = "ppb",
  token = "1234",
  query = "PPBstats_data_agro"
  )
```

and then transform if for `PPBstats` format.

```r
library(PPBstats)
data = PPBstats::format_data_PPBstats(data_agro, type = "data_agro")
```





NEWS.md
# shinemas2R 0.1

- first commit with core of the package, cf #1
- website structure and first text


R/shinemas.R
#' Query the database SHiNeMaS throught an API and format data for specific R packages
#'
#' @description
#' \code{shinemas} queries the database SHiNeMaS throught an API and formats data for specific R packages
#' 
#' @param db_url url of the database
#' 
#' @param user user name
#' 
#' @param password password
#' 
#' @param token token to connect to the database
#' 
#' 
#' @param query_type name of the query. Possible values are
#' \itemize{
#'  \item PPBstats_data_network_unipart_seed_lots
#'  \item PPBstats_data_agro
#'  \item PPBstats_data_agro_SR
#'  \item PPBstats_data_agro_HA
#' }
#' 
#' @param specie filter: specie to keep
#' 
#' @param project filter: project to keep
#' 
#' @param variable filter: variable to keep
#' 
#' @param germplasm filter: germplasm to keep
#' 
#' @param germplasm_type filter: germplasm type to keep
#' 
#' @param relation_type filter: relation type to keep
#' 
#' @param location filter: location to keep
#' 
#' @param year filter: year to keep
#' 
#' @return 
#' The function return a dataframe formated for specific query type and packages
#'  
#' @details 
#' \itemize{
#'  \item \strong{SHiNeMaS}: The database SHiNeMaS (Seed History and Network Management System) is available here: \url{https://sourcesup.renater.fr/frs/?group_id=2295}.
#'  The query is based on a webservice developped within SHiNeMaS v2 by Laetitia Courgey and Yannick de Oliviera.
#'  The token can be taken through the web interface of SHiNeMaS by cliking "Get Token" on the top left side of the page.
#'  \item \strong{query_type}: The query_type is under the format [package-name]_[query-name].
#'  For example PPBstats_data_agro stands for data_agro format for package PPBstats.
#'  \item \strong{Package supported}:
#'  \itemize{
#'   \item \pkg{PPBstats}: PPBstats is an  R package for Participatory Plant Breeding statisticial analyses.
#'   More information are available here: \url{https://priviere.github.io/PPBstats_web_site/}
#'   \itemize{
#'    \item For query_type = data_agro and data_agro_HA 
#'     \itemize{
#'     \item filter year refers to sown year of seed-lot parent or year of selection of seed-lot parent
#'     \item relation_type is either reproduction or selection
#'     }
#'   }
#'  }
#' }
#' 
#' @examples 
#' \dontrun{
#' data_agro = shinemas2R::shinemas(
#'   db_url = "shinemas.local.seed",
#'   user = "wheat",
#'   password = "ppb",
#'   token = "1234",
#'   query_type = "PPBstats_data_agro"
#' )
#' 
#' library(PPBstats)
#' data = format_data_PPBstats(data_agro, type = "data_agro")
#' }
#' 
#' @author Pierre Riviere
#' 
#' @import httr
#' @import jsonlite
#' 
#' @export
#' 
shinemas = function(
  db_url = "shinemas.local.seed",
  user = "wheat",
  password = "ppb",
  token = "1234",
  query_type = "PPBstats_data_agro",
  specie = NULL,
  project = NULL,
  variable = NULL,
  germplasm = NULL,
  germplasm_type = NULL,
  relation_type = NULL,
  location = NULL,
  year = NULL
  ){
  
  match.arg(query_type, c(
    "PPBstats_data_network_unipart_seed_lots", 
    "PPBstats_data_agro", 
    "PPBstats_data_SR", 
    "PPBstats_data_HA")
    )
  
  get_data_from_shinemas = function(db_url, user, password, token, query){
    url_shinemas = paste(db_url, "/wsR/",query,"?token=", token, sep = "")
    data_shinemas = httr::GET(url_shinemas, authenticate(user, password))
    data = jsonlite::fromJSON(data_shinemas) 
    return(data)
    }
  
  # query_type == "PPBstats_data_network_unipart_seed_lots" ----------
  if( query_type == "PPBstats_data_network_unipart_seed_lots"){
    
    data = get_data_from_shinemas(db_url, user, password, token, query = "data_network_unipart_seed_lots")
    
    vec_fac = c("specie", "project", "seed_lot_parent", "seed_lot_child", "relation_type", "relation_year_start", "relation_year_end", "germplasm_parent", 
                "location_parent", "year_parent", "germplasm_child", "location_child", "year_child", "species_parent", "project_parent", "germplasm_type_parent", 
                "comments_parent", "species_child", "project_child", "germplasm_type_child", "comments_child")
    for(v in vec_fac){ data[v] = as.factor(as.character(v)) }
    
    vec_num = c("long_parent", "lat_parent", "long_child", "lat_child", 
                "alt_parent", "total_generation_nb_parent", "local_generation_nb_parent", "generation_confidence_parent", 
                "alt_child", "total_generation_nb_chill", "local_generation_nb_child", "generation_confidence_child")
    for(v in vec_num){ data[v] = as.numeric(as.character(v)) }
    
  }
  
  # query_type == "PPBstats_data_agro" ----------
  if( query_type == "PPBstats_data_agro"){
    
    data = get_data_from_shinemas(db_url, user, password, token, query = "data_agro")
    
    vec_fac = c("specie", "project", "seed_lot", "location", "year", "germplasm", "block", "X", "Y")
    for(v in vec_fac){ data[v] = as.factor(as.character(v)) }
    
    vec_num = c("lat", "long")
    for(v in vec_num){ data[v] = as.numeric(as.character(v)) }
    
  }
  
  # query_type == "PPBstats_data_SR" ----------
  if( query_type == "PPBstats_data_agro_SR"){
    
    data = get_data_from_shinemas(db_url, user, password, token, query = "data_agro_SR")
    
    vec_fac = c("specie", "project", "seed_lot", "location", "year", "germplasm", "block", "X", "Y", "expe_id", "group", "version")
    for(v in vec_fac){ data[v] = as.factor(as.character(v)) }
    
    vec_num = c("lat", "long")
    for(v in vec_num){ data[v] = as.numeric(as.character(v)) }
    
  }
  
  # query_type == "PPBstats_data_HA" ----------
  if( query_type == "PPBstats_data_agro_HA"){
    
    data = get_data_from_shinemas(db_url, user, password, token, query = "data_agro_HA")
    
    vec_fac = c("specie", "project", "seed_lot", "location", "year", "germplasm", "block", "X", "Y", "origin", "version")
    for(v in vec_fac){ data[v] = as.factor(as.character(v)) }
    
    vec_num = c("lat", "long")
    for(v in vec_num){ data[v] = as.numeric(as.character(v)) }
    
  }
  
  return(data)
}
